torch/csrc/jit/codegen/cuda/lower_double_buffer.cpp

```diff
@@ -279,7 +279,13 @@ class DoubleBufferLoopCloner : public kir::IrVisitor {
          is_double_buffer_load_expr) ||
         (loop_type_ == DoubleBufferLoopStage::Epilog &&
          !is_double_buffer_load_expr)) {
-      cloned_scopes_.back()->push_back(expr);
+      if (supportInlinePredicate(expr) && expr->isA<LoadStoreOp>()) {
+        auto ldst = expr->as<LoadStoreOp>();
+        cloned_scopes_.back()->push_back(IrBuilder::create<LoadStoreOp>(
+            ldst->opType(), ldst->out(), ldst->in()));
```

Comment:
Well, at this point, I think `expr` should have no predicate (i.e., `expr->predicate() == null`) as embedding its predicate into the IR node is done at the unroll pass, which comes after the double buffering pass. So, I don't see any reason we would need to have this conditional processing.

Reply:
OK, I found why creating a new instance of `LoadStoreOp` is necessary here. This part of the code creates a copy of a loop nest, and we want to have different predicates for different copies. If a new instance is not created, a single `LoadStoreOp` is shared between different copies of a loop nest. That's also the case with any other type of expressions, but for `LoadStoreOp`, we embed a predicate in the instance itself with `setPredicate()`, so all loop nests sharing the same `LoadStoreOp` would end up having the same predicate, which isn't what we want. For any other type of expressions, we add `IfThenElse` around each expression after this double buffering pass, so expression sharing is not an issue.

This is a bit tricky to make it right and consistent. It seems to me that the right approach would be: 1) Make expression instances non-modifiable; and 2) Clone an expression and set its predicate in `UnrollPass`. Expression IR nodes are generally non-modifiable already, and I believe that's generally the design baseline, but `setPredicate` and `setWritePredicate` violate it as they modify embedded predicates. I think removing them or making them `private` would make the overall IR design more consistent. Instead of using those methods, we should use the constructors to set predicates.

It would require some changes in the unroll pass, e.g., https://github.com/csarofeen/pytorch/blob/devel/torch/csrc/jit/codegen/cuda/lower_unroll.cpp#L106, but I think it'd be more consistent overall.

Reply:
WIP at https://github.com/csarofeen/pytorch/pull/2025